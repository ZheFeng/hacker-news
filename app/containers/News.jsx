import React, { Component, PropTypes } from 'react';
import { List, fromJS } from 'immutable';

import ContentList from '../components/ContentList';
import Pagination from '../components/Pagination';


function fetchItem(id) {
  const newsApi = `https://hacker-news.firebaseio.com/v0/item/${id}.json`;
  return fetch(newsApi)
  .then(response => response.json())
  .then(data => fromJS(data));
}

function fetchData(start, numberPerPage, ids) {
  const news = ids.slice(start, start + numberPerPage);
  return Promise.all(news.map(fetchItem))
  .then(data => fromJS(data));
}
function fetchIds(topic) {
  const newsApi = `https://hacker-news.firebaseio.com/v0/${topic}.json`;
  return fetch(newsApi)
  .then(response => response.json())
  .then(data => fromJS(data));
}

class News extends Component {
  static displayName = 'News';
  static propTypes = {
    topic: PropTypes.string,
  }
  static defaultProps = {
    topic: 'topstories',
  }
  state = {
    data: new List(),
    ids: new List(),
    start: 0,
    numberPerPage: 10,
    fetching: true,
  }
  componentWillMount() {
    const { topic } = this.props;
    const { numberPerPage } = this.state;
    fetchIds(topic)
    .then(ids => this.setState({ ids }) || ids)
    .then(fetchData.bind(this, 0, numberPerPage))
    .then(data => this.setState({ data, fetching: false }));
  }
  componentWillReceiveProps(nextProps) {
    const { topic } = nextProps;
    const { numberPerPage } = this.state;
    if (this.props.topic !== topic) {
      this.cleanup();
      fetchIds(topic)
      .then(ids => this.setState({ ids }) || ids)
      .then(fetchData.bind(null, 0, numberPerPage))
      .then(data => this.setState({ data, fetching: false }));
    }
  }
  onPageChange(index) {
    const { numberPerPage } = this.state;
    const start = (index - 1) * numberPerPage;
    if (start !== this.state.start) {
      this.setState({ start, fetching: true, data: new List() });
      fetchData(start, numberPerPage, this.state.ids)
      .then(data => this.setState({ data, fetching: false }));
    }
  }
  render() {
    const { start, numberPerPage, ids, fetching } = this.state;
    const total = Math.ceil(ids.size / numberPerPage);
    const current = (start / numberPerPage) + 1;
    const onPageChange = this::this.onPageChange;
    const node = fetching ? (
      <div>Loading</div>
    ) : (<ContentList data={this.state.data} />);
    return (
      <div>
        <Pagination
          total={total || 1}
          current={current}
          onPageChange={onPageChange}
          disabled={fetching}
        />
        {node}
      </div>
    );
  }
  cleanup() {
    this.setState({
      data: new List(),
      ids: new List(),
      start: 0,
      numberPerPage: 10,
      fetching: true,
    });
  }
}

export default News;
