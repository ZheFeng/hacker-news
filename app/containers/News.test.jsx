/*
  eslint-disable
  import/first, import/no-extraneous-dependencies, max-nested-callbacks
*/
/* global jest, describe, it, expect */
import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { createStore } from 'redux';

import Store from '../Store';

const store = createStore(Store);

import News from './News';

describe('<News />', () => {
  it('Snapshot Test', () => {
    const component = renderer.create(
      <Provider store={store}><News /></Provider>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
