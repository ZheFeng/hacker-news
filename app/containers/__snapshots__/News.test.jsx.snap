exports[`<News /> Snapshot Test 1`] = `
<div>
  09:53:49 GMT+1100 (AEDT)
  <div>
    Loading...
  </div>
</div>
`;
